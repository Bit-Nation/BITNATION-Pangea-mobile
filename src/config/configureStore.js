// @flow

import { createStore, applyMiddleware, compose, type Store } from 'redux';
import logger from 'redux-logger';
import createSagaMiddleware from 'redux-saga';
import createSagaMonitor from '@clarketm/saga-monitor';
import reducers from '../reducers';
import rootSaga from '../sagas';

/**
 * @desc Configures a Redux store.
 * @return {Store} Created store object.
 */
export default function configureStore(): Store {
  const sagaMonitor = createSagaMonitor({
    level: 'log',
    actionDispatch: true,
  });
  const sagaMiddleware = createSagaMiddleware({ sagaMonitor });

  const enhancer = compose(
    applyMiddleware(sagaMiddleware),
    applyMiddleware(logger),
  );

  const store = createStore(reducers, enhancer);
  sagaMiddleware.run(rootSaga);
  return store;
}
