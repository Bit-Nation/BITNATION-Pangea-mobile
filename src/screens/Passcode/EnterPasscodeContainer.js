// @flow

import React from 'react';
import { connect } from 'react-redux';
import { View } from 'react-native';
import PinCodeScreen from './PinCode/index';
import { type State as SettingsState } from '../../reducers/settings';
import NavigatorComponent from '../../components/common/NavigatorComponent';
import i18n from '../../global/i18n';
import { checkPinCode, checkPassword, login } from '../../actions/accounts';
import { errorAlert } from '../../global/alerts';
import PasswordScreen from './Password/index';
import styles from './PinCode/styles';
import BackgroundImage from '../../components/common/BackgroundImage';
import FakeNavigationBar from '../../components/common/FakeNavigationBar';
import ScreenTitle from '../../components/common/ScreenTitle';
import type { AsyncTask } from '../../utils/asyncTask';
import Loading from '../../components/common/Loading';

type Props = {
  /**
   * @desc Callback on cancellation of entering passcode.
   */
  onCancel: () => void,
  /**
   * @desc Callback on successful passcode entering.
   */
  onSuccess: () => void,
  /**
   * @desc Function that is called when user press on forget password button.
   */
  onForget: () => void,
  /**
   * @desc Title of the screen.
   */
  title: string,
  /**
   * @desc Id of account which passcode will be entered.
   */
  accountId: string,
  /**
   * @desc Flag if there is currently logged in account.
   */
  isLoggedIn: boolean,
  /**
   * @desc Task that describes current login status.
   */
  loginTask: AsyncTask<void>,
}

type Actions = {
  /**
   * @desc Check entered pin code.
   */
  checkPinCode: (pinCode: string, accountId: string, callback: (success: boolean) => void) => void,
  /**
   * @desc Check entered password.
   */
  checkPassword: (password: string, accountId: string, callback: (success: boolean) => void) => void,
  /**
   * @desc Action to perform a login.
   */
  login: (password: string, accountId: string) => void,
}

class EnterPasscodeContainer extends NavigatorComponent<Props & Actions & SettingsState> {
  static defaultProps;

  componentDidUpdate(prevProps: Props & Actions & SettingsState) {
    if (prevProps.loginTask !== this.props.loginTask && this.props.loginTask.error != null) {
      errorAlert(this.props.loginTask.error);
    }
  }

  onCheckFinished = (success: boolean) => {
    if (success === true) {
      this.props.onSuccess();
      return;
    }

    if (this.props.passcodeType.type === 'pinCode') {
      errorAlert(i18n.t('error.invalidPinCode'));
    } else {
      errorAlert(i18n.t('error.invalidPassword'));
    }
  };

  onPasscodeEntered = (passcode) => {
    const { accountId, isLoggedIn } = this.props;
    if (isLoggedIn === false) {
      this.props.login(accountId, passcode);
      return;
    }

    if (this.props.passcodeType.type === 'pinCode') {
      this.props.checkPinCode(passcode, accountId, this.onCheckFinished);
    } else {
      this.props.checkPassword(passcode, accountId, this.onCheckFinished);
    }
  };

  onForgetPasscode = () => {
    this.props.onForget();
  }

  renderPasscodeScreen() {
    const {
      navigator, passcodeType, title, onCancel, isLoggedIn,
    } = this.props;
    const shouldShowForget = isLoggedIn === false;
    if (passcodeType.type === 'pinCode') {
      return (<PinCodeScreen
        navigator={navigator}
        pinCodeLength={passcodeType.length}
        instruction={i18n.t('screens.pinCode.enterInstruction')}
        title={title}
        shouldShowCancel
        onCancel={onCancel}
        onSubmit={this.onPasscodeEntered}
        shouldShowForget={shouldShowForget}
        onForget={this.onForgetPasscode}
      />);
    }

    return (<PasswordScreen
      navigator={navigator}
      instruction={i18n.t('screens.password.enterInstruction')}
      title={title}
      shouldShowCancel
      onCancel={onCancel}
      onSubmit={this.onPasscodeEntered}
      shouldShowForget={shouldShowForget}
      onForget={this.onForgetPasscode}
    />);
  }

  render() {
    const { title } = this.props;
    return (
      <View
        testID='enterPasscode_wrapperView'
        style={styles.screenContainer}
      >
        <BackgroundImage />
        <FakeNavigationBar />
        <ScreenTitle title={title} />
        {this.renderPasscodeScreen()}
        {this.props.loginTask.inProgress === true && <Loading />}
      </View>
    );
  }
}

EnterPasscodeContainer.defaultProps = {
  onCancel: () => undefined,
  onSuccess: () => undefined,
  title: '',
};

const mapStateToProps = state => ({
  ...state.settings,
  loginTask: state.accounts.login,
  isLoggedIn: state.accounts.currentAccountId !== null,
});

const mapDispatchToProps = dispatch => ({
  checkPinCode(pinCode, accountId, callback) {
    dispatch(checkPinCode(pinCode, accountId, callback));
  },
  checkPassword(password, accountId, callback) {
    dispatch(checkPassword(password, accountId, callback));
  },
  login(password, accountId) {
    dispatch(login(password, accountId));
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(EnterPasscodeContainer);
