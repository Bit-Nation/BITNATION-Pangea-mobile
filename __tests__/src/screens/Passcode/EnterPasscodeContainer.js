import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';

import EnterPasscodeContainer from '../../../../src/screens/Passcode/EnterPasscodeContainer';

describe('EnterPasscodeContainer', () => {
  let enterPasscodeContainer;

  const initialStateMock = {
    accounts: {
      login: {
        error: 'Invalid',
      },
    },
    settings: {
      passcodeType: {
        type: 'pinCode',
        length: 6,
      },
    },
  };
  const storeMock = configureStore([]);
  const propsMock = {
    navigator: {
      push: jest.fn(),
      pop: jest.fn(),
      setOnNavigatorEvent: jest.fn(),
    },
    checkPinCode: jest.fn(),
    checkPassword: jest.fn(),
    login: jest.fn(),
    onSuccess: jest.fn(),
    onCancel: jest.fn(),
  };

  beforeEach(() => {
    enterPasscodeContainer = shallow(<EnterPasscodeContainer {...propsMock} store={storeMock(initialStateMock)} />);
  });

  test('Rendering', () => {
    expect(enterPasscodeContainer).toMatchSnapshot();
  });

  test('check required props', () => {
    expect(enterPasscodeContainer.props().navigator).toBeDefined();
  });
});
