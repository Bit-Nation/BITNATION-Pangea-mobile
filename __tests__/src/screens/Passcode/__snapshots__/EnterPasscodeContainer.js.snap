// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnterPasscodeContainer Rendering 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(EnterPasscodeContainer)
    checkPassword={[MockFunction]}
    checkPinCode={[MockFunction]}
    login={[MockFunction]}
    navigator={
      Object {
        "pop": [MockFunction],
        "push": [MockFunction],
        "setButtons": [MockFunction],
        "setOnNavigatorEvent": [MockFunction],
      }
    }
    onCancel={[MockFunction]}
    onSuccess={[MockFunction]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "cancelLogin": [Function],
      "checkPassword": [Function],
      "checkPinCode": [Function],
      "isLoggedIn": true,
      "login": [Function],
      "loginTask": Object {
        "error": "Invalid",
      },
      "navigator": Object {
        "pop": [MockFunction],
        "push": [MockFunction],
        "setButtons": [MockFunction],
        "setOnNavigatorEvent": [MockFunction],
      },
      "onCancel": [MockFunction],
      "onSuccess": [MockFunction],
      "passcodeType": Object {
        "length": 6,
        "type": "pinCode",
      },
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "storeSubscription": Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
      "title": "",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "cancelLogin": [Function],
        "checkPassword": [Function],
        "checkPinCode": [Function],
        "isLoggedIn": true,
        "login": [Function],
        "loginTask": Object {
          "error": "Invalid",
        },
        "navigator": Object {
          "pop": [MockFunction],
          "push": [MockFunction],
          "setButtons": [MockFunction],
          "setOnNavigatorEvent": [MockFunction],
        },
        "onCancel": [MockFunction],
        "onSuccess": [MockFunction],
        "passcodeType": Object {
          "length": 6,
          "type": "pinCode",
        },
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "storeSubscription": Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": [Function],
        },
        "title": "",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
